### PR TITLE
Support SSL on the file staging server

### PR DESCRIPTION
**Pull request by mccheah**
Allow providing a keyStore or key+certificate PEM files to have the staging server listen over TLS. 

Note that we're also preemptively allowing the server's configuration to be provided in a properties file. This is to allow properties to be provided through a ConfigMap more easily if this file staging server were to be deployed in Kubernetes. The alternative would be to specify all of these properties as system properties or command-line arguments, but I think our previous work on similar constructs has shown that parsing out command line arguments and expecting them to be set correctly in the Dockerfile can be tedious.

In the case that the user prefer not to have the key and keystore passwords be provided through the ConfigMap - e.g. they would prefer to mount K8s secrets - the key and keystore password values can be provided in files as well.

### CODE REVIEW

**Comment by mccheah** [comment]
Note that this is merging into #220 , which in turn is merging into #212. We should:

1. Merge #212 first
2. Retarget #220 to `branch-2.1-kubernetes`
3. Merge #220 
4. Retarget this PR to `branch-2.1-kubernetes`
5. Merge this

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala [hunk 1]
```diff
@@ -288,4 +288,37 @@ package object config {
...
+      .intConf
+      .createWithDefault(10000)
+
+  private[spark] val DEPENDENCY_SERVER_KEY_PEM =
+    ConfigBuilder("spark.ssl.kubernetes.dependencyserver.keyPem")
+      .doc("Key PEM file to use when having the Kubernetes dependency server listen on TLS.")
+      .stringConf
+      .createOptional
+
+  private[spark] val DEPENDENCY_SERVER_CERT_PEM =
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala]: put all these new config options into the docs file.  Also needs to include what of `file://` `local://` `hdfs://` protocols are supported here

Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala]: I was planning to save all documentation for a final sweep where we eliminate all of the v1 submission logic and replace it with v2. At that point the docs should be updated to remove the old options and replace them with the new ones.

**Comment by mccheah** [comment]
rerun integration tests please

**Comment by mccheah** [comment]
rerun integration test

**Comment by mccheah** [comment]
rerun integration tests please

**Comment by mccheah** [comment]
rerun integration test please

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/v2/ResourceStagingServer.scala [hunk 2]
```diff
@@ -45,17 +57,71 @@ private[spark] class ResourceStagingServer(
...
+      // Call this full constructor to set this, which forces daemon threads:
+      new ScheduledExecutorScheduler("DependencyServer-Executor", true),
+      null,
+      -1,
+      -1,
+      resolvedConnectionFactories: _*)
     connector.setPort(port)
     server.addConnector(connector)
     server.setHandler(contextHandler)
     server.start()
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/v2/ResourceStagingServer.scala:85]: does this log some kind of "server started on hostname:port" message?  Want to have a clear "I'm ready" message to look for in server logs

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/v2/ResourceStagingServerSslOptionsProvider.scala [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+import org.apache.spark.{SecurityManager => SparkSecurityManager, SparkConf, SparkException, SSLOptions}
+import org.apache.spark.deploy.kubernetes.config._
+import org.apache.spark.deploy.rest.kubernetes.v1.PemsToKeyStoreConverter
+
+private[spark] trait ResourceStagingServerSslOptionsProvider {
+  def getSslOptions: SSLOptions
+}
+
+private[spark] class ResourceStagingServerSslOptionsProviderImpl(sparkConf: SparkConf)
+    extends ResourceStagingServerSslOptionsProvider {
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/v2/ResourceStagingServerSslOptionsProvider.scala]: I find a server logging all its non-sensitive config on startup helps with debugging later -- maybe DEBUG log everything in SparkConf and make sure DEBUG is enabled for this class?

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/v2/ResourceStagingServerSslOptionsProvider.scala [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    val maybeKeyPasswordFile = sparkConf.get(DEPENDENCY_SERVER_KEYSTORE_KEY_PASSWORD_FILE)
+    requireNandDefined(baseSslOptions.keyStore, maybeKeyPem,
+      "Shouldn't provide both key PEM and keyStore files for TLS.")
+    requireNandDefined(baseSslOptions.keyStore, maybeCertPem,
+      "Shouldn't provide both certificate PEM and keyStore files for TLS.")
+    requireNandDefined(baseSslOptions.keyStorePassword, maybeKeyStorePasswordFile,
+      "Shouldn't provide both the keyStore password value and the keyStore password file.")
+    requireNandDefined(baseSslOptions.keyPassword, maybeKeyPasswordFile,
+      "Shouldn't provide both the keyStore key password value and the keyStore key password file.")
+    maybeKeyPem.foreach { _ =>
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/v2/ResourceStagingServerSslOptionsProvider.scala]: refactor to a `requireSecondIfFirst` method

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/rest/kubernetes/v2/ResourceStagingServerSslOptionsProviderSuite.scala [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+import org.scalatest.BeforeAndAfter
+
+import org.apache.spark.{SparkConf, SparkFunSuite, SSLOptions}
+import org.apache.spark.deploy.kubernetes.SSLUtils
+import org.apache.spark.util.Utils
+
+class ResourceStagingServerSslOptionsProviderSuite extends SparkFunSuite with BeforeAndAfter {
+
+  private val SSL_TEMP_DIR = Utils.createTempDir(namePrefix = "resource-staging-server-ssl-test")
+  private val KEYSTORE_FILE = new File(SSL_TEMP_DIR, "keyStore.jks")
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/rest/kubernetes/v2/ResourceStagingServerSslOptionsProviderSuite.scala]: convention is all-lowercase `keystore.jks` I think

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/rest/kubernetes/v2/ResourceStagingServerSslOptionsProviderSuite.scala [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+  private var sparkConf: SparkConf = _
+  private var sslOptionsProvider: ResourceStagingServerSslOptionsProvider = _
+
+  before {
+    sparkConf = new SparkConf(true)
+    sslOptionsProvider = new ResourceStagingServerSslOptionsProviderImpl(sparkConf)
+  }
+
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/rest/kubernetes/v2/ResourceStagingServerSslOptionsProviderSuite.scala]: clear the `SSL_TEMP_DIR` between tests to reduce risk of polluting other test runs

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/rest/kubernetes/v2/ResourceStagingServerSuite.scala [hunk 7]
```diff
@@ -39,30 +40,50 @@ import org.apache.spark.util.Utils
...
+  test("Accept file and jar uploads and downloads") {
     server.start()
+    runUploadAndDownload(SSLOptions())
   }
 
-  override def afterAll(): Unit = {
-    server.stop()
+  test("Enable SSL on the server") {
+    val (keyStore, trustStore) = SSLUtils.generateKeyStoreTrustStorePair(
+      "127.0.0.1", "changeit", "changeit", "changeit")
```
Comment (ash211) [inline resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/rest/kubernetes/v2/ResourceStagingServerSuite.scala]: make these 3 values different so it's clear what's what and make sure the values a couple lines down are lined up

**Review by ash211 — COMMENTED** [review]
Very nicely tested!  Some minor suggestions.  The biggest thing is probably logging and the overall logging story -- do you want to tackle that here or in a separate PR?

When running some manual tests I think we're going to find logging invaluable

File: resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/rest/kubernetes/v2/ResourceStagingServerSslOptionsProviderSuite.scala (hunk 5 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/rest/kubernetes/v2/ResourceStagingServerSslOptionsProviderSuite.scala]: The convention is to capitalize the "S" on keyStore.

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/v2/ResourceStagingServer.scala (hunk 2 above)
Comment (mccheah) [reply resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/v2/ResourceStagingServer.scala:85]: I don't think we have the hostname here but Spark tends to log "Started server listening on port X", and we should do something similar here.

**Review by ash211 — APPROVED** [review]
LGTM, will merge later today after https://github.com/apache-spark-on-k8s/spark/pull/220

**Comment by ash211** [comment]
rerun unit tests please

**Pull request merged**